### PR TITLE
Support plugin-only secrets

### LINKED ISSUE #1071: [Important] Limit secret access to plugins only
similar to the images filter for secrets, add a checkbox to only inject secret if step is exec as plugin.

this ensure only the intended entry-point do get the secrets to handle.

this does help if a plugin is not based from scratch image but do contain an shell that could be used.

---
bounty: 50$

### PR DESCRIPTION
**Pull request by qwerty287**
Closes #1071

### CODE REVIEW

File: pipeline/frontend/yaml/compiler/convert.go [hunk 1]
```diff
@@ -192,3 +192,7 @@ func (c *Compiler) stepWorkdir(container *yaml.Container) string {
 	}
 	return filepath.Join(c.base, c.path, container.Directory)
 }
+
+func isPlugin(container *yaml.Container) bool {
+	return len(container.Commands) < 1 && len(container.Command) < 1
```
Comment (lafriks) [inline pipeline/frontend/yaml/compiler/convert.go:197]: This is more readable imho
```suggestion
	return len(container.Commands) == 0 && len(container.Command) == 0
```

File: pipeline/frontend/yaml/compiler/compiler.go [hunk 2]
```diff
@@ -35,9 +35,18 @@ type Registry struct {
...
+	Value      string
+	Match      []string
+	PluginOnly bool
+}
+
+func (s *Secret) Available(container *yaml.Container) bool {
+	return (len(s.Match) == 0 || matchImage(container.Image, s.Match...)) && (!s.PluginOnly || isPlugin(container))
+}
+
+func isPlugin(container *yaml.Container) bool {
```
Comment (6543) [inline pipeline/frontend/yaml/compiler/compiler.go:48]: can we move this func as public func for the strict Container ?

File: web/src/assets/locales/en.json [hunk 3]
```diff
@@ -258,6 +259,7 @@
                     "images": "Available for following images",
                     "desc": "Comma separated list of images where this secret is available, leave empty to allow all images"
                 },
+                "plugins_only": "Only available for plugins",
```
Comment (6543) [inline web/src/assets/locales/en.json:262]: I disslike the redundancy  - why was it added in the first place?

File: server/model/secret.go [hunk 4]
```diff
@@ -68,15 +68,16 @@ type SecretStore interface {
...
-	Events     []WebhookEvent `json:"event"           xorm:"json 'secret_events'"`
-	SkipVerify bool           `json:"-"               xorm:"secret_skip_verify"`
-	Conceal    bool           `json:"-"               xorm:"secret_conceal"`
+	ID          int64          `json:"id"              xorm:"pk autoincr 'secret_id'"`
+	Owner       string         `json:"-"               xorm:"NOT NULL DEFAULT '' UNIQUE(s) INDEX 'secret_owner'"`
+	RepoID      int64          `json:"-"               xorm:"NOT NULL DEFAULT 0 UNIQUE(s) INDEX 'secret_repo_id'"`
+	Name        string         `json:"name"            xorm:"NOT NULL UNIQUE(s) INDEX 'secret_name'"`
+	Value       string         `json:"value,omitempty" xorm:"TEXT 'secret_value'"`
+	Images      []string       `json:"image"           xorm:"json 'secret_images'"`
+	PluginsOnly bool           `json:"plugins_only"           xorm:"secret_plugins_only"`
```
Comment (6543) [inline server/model/secret.go:77]: intendation

**Comment by 6543** [comment]
```diff
diff --git a/web/src/assets/locales/en.json b/web/src/assets/locales/en.json
index ce65ab16e..87f770fff 100644
--- a/web/src/assets/locales/en.json
+++ b/web/src/assets/locales/en.json
@@ -27,6 +27,11 @@
         "sec_short": "sec",
         "not_started": "not started yet"
     },
+    "settings": {
+      "secrets": {
+          "plugins_only": "Only available for plugins"
+      }
+    },
     "repo": {
         "manual_pipeline": {
           "title": "Trigger a manual pipeline run",
@@ -121,7 +126,6 @@
                     "events": "Available at following events",
                     "pr_warning": "Please be careful with this option as a bad actor can submit a malicious pull request that exposes your secrets."
                 },
-                "plugins_only": "Only available for plugins",
                 "edit": "Edit secret",
                 "delete":"Delete secret"
             },
@@ -259,7 +263,6 @@
                     "images": "Available for following images",
                     "desc": "Comma separated list of images where this secret is available, leave empty to allow all images"
                 },
-                "plugins_only": "Only available for plugins",
                 "events": {
                     "events": "Available at following events",
                     "pr_warning": "Please be careful with this option as a bad actor can submit a malicious pull request that exposes your secrets."
@@ -288,7 +291,6 @@
                     "images": "Available for following images",
                     "desc": "Comma separated list of images where this secret is available, leave empty to allow all images"
                 },
-                "plugins_only": "Only available for plugins",
                 "events": {
                     "events": "Available at following events",
                     "pr_warning": "Please be careful with this option as a bad actor can submit a malicious pull request that exposes your secrets."
diff --git a/web/src/components/secrets/SecretEdit.vue b/web/src/components/secrets/SecretEdit.vue
index 8bff336bc..bb6d9a4f1 100644
--- a/web/src/components/secrets/SecretEdit.vue
+++ b/web/src/components/secrets/SecretEdit.vue
@@ -17,7 +17,7 @@
       <InputField :label="$t(i18nPrefix + 'images.images')">
         <TextField v-model="images" :placeholder="$t(i18nPrefix + 'images.desc')" />
 
-        <Checkbox v-model="innerValue.plugins_only" class="mt-4" :label="$t(i18nPrefix + 'plugins_only')" />
+        <Checkbox v-model="innerValue.plugins_only" class="mt-4" :label="$t('settings.secrets.plugins_only')" />
       </InputField>
 
       <InputField :label="$t(i18nPrefix + 'events.events')">
```

so we have a general "settings" section and only if it differs between org/admin/repo use a specific prefix ... ?

**Comment by 6543** [comment]
the redundant translation entrys is unrelated to this but need to be solved ... in an other pull

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
@qwerty287 please submit a ref to the issue to OpenCollective for 50usd

**Comment by 6543** [comment]
@qwerty287 :

https://opencollective.com/woodpecker-ci/expenses/new
image

**Comment by 6543** [comment]
you just need to create an account and use that in "who is paied" ...

**Comment by 6543** [comment]
... only the admins (anbraten & me) + OpenSourceCollective will see that info

**Comment by 6543** [comment]
@qwerty287 DM -> discord  (in general we need a way to pay people who did a bounty less complicated!)

**Pull request merged**